### PR TITLE
Support Dream-v0-Base-7B

### PR DESCRIPTION
**Pull request by Jayon02**
## Motivation

Support Dream-org/Dream-v0-Base-7B

## Modifications

- Add initial Dream support to the dLLM runtime.
- Add full-prefill generation and corresponding scheduler support for Dream.
- Preserve the existing block-diffusion path for LLaDA models.
- Add unit test.

## Accuracy Tests

Model | Benchmark | Examples | Max Tokens | Accuracy 
-- | -- | -- | -- | -- 
Dream-v0-Base-7B | MMLU | 1,000 | 128 | 43.0% 

## Speed Tests and Profiling

Model | Concurrency | Requests | Req Throughput | Output Throughput | Total Throughput | Mean E2E
-- | -- | -- | -- | -- | -- | --
Dream-v0-Base-7B | 4 | 20 | 0.13 req/s | 7.16 tok/s | 16.62 tok/s | 31.42 s

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35107201860](https://github.com/sgl-project/sglang/actions/runs/35107201860)
Latest PR Test (Extra): :x: [Run #35107200935](https://github.com/sgl-project/sglang/actions/runs/35107200935)
Latest PR Test (AMD ROCm 10): :x: [Run #35107202145](https://github.com/sgl-project/sglang/actions/runs/35107202145)

### CODE REVIEW

**Comment by Jayon02** [comment]
@ClawSeven @merrymercy 
This PR adds support for Dream-org/Dream-v0-Base-7B. Could you please help review it and enable CI?

File: python/sglang/srt/dllm/algorithm/base.py [hunk 1]
```diff
@@ -64,26 +66,63 @@ def run(
...
+        else:
+            start_list = self._block_start_list(forward_batch)
 
         out = model_runner.forward(forward_batch, pp_proxy_tensors=None)
         # No mask to denoise: return empty so process_batch_result_dllm skips the
         # stream branch (matches the pre-refactor behavior).
-        if all(start == self.block_size for start in start_list):
-            return out.logits_output, [], None, None, out.can_run_graph
+        if all(start == length for start, length in zip(start_list, sequence_lengths)):
+            return out.logits_output, [], None, None, out.can_run_graph, None
```
Comment (rwang5203) [inline python/sglang/srt/dllm/algorithm/base.py]: With `--no-dllm-fdfo`, zero-token Dream requests skip completion and KV release, leaving requests stuck. Could you try finishing empty canvases and releasing their KV?

File: test/registered/unit/models/test_dream.py [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+    def _make_server_args_for_graph_test(self, disable_cuda_graph):
+        args = ServerArgs.__new__(ServerArgs)
+        args.dllm_algorithm = "Dream"
+        args.dllm_fdfo = True
+        args.tp_size = 1
+        args.pp_size = 1
+        args.disable_radix_cache = False
+        args.disable_cuda_graph = disable_cuda_graph
+        args.cuda_graph_config = SimpleNamespace(
+            decode=SimpleNamespace(
```
Comment (rwang5203) [inline test/registered/unit/models/test_dream.py]: The graph-enabled guard test raises `TypeError` because `with_phase()` receives `SimpleNamespace` objects. Maybe try building the fixture with `CudaGraphConfig` and `PhaseConfig`?

File: python/sglang/srt/dllm/mixin/scheduler.py [hunk 3]
```diff
@@ -75,19 +75,57 @@ def process_batch_result_dllm(
...
+                        self.metrics_reporter.num_generated_tokens += len(
+                            req.output_ids
+                        )
+                        if req.output_ids:
+                            req.update_finish_state(
+                                new_accepted_len=len(req.output_ids)
+                            )
+                        else:
+                            req.finished_reason = FINISH_LENGTH(length=0)
+                            req.finished_len = 0
```
Comment (rwang5203) [inline python/sglang/srt/dllm/mixin/scheduler.py]: For empty Dream canvases, this branch ignores queued aborts and reports successful length completion. One option is to preserve `to_finish` through the normal finish-state update.

**Comment by Jayon02** [comment]
Thanks @rwang5203. I’ve addressed all of your comments and updated the relevant tests.

**Comment by rwang5203** [comment]
The graph dataclasses are fixed, but both guard cases raise TypeError at [ServerArgs.__new__(ServerArgs)](https://github.com/sgl-project/sglang/blob/63ccbf7a737912be4035b0286a34924756d779e7/test/registered/unit/models/test_dream.py#L309). Could you construct the fixture with ServerArgs(model_path="dummy")?

File: python/sglang/srt/sampling/sampling_params.py [hunk 4]
```diff
@@ -158,6 +158,8 @@ class SamplingParams(msgspec.Struct, kw_only=True, array_like=True):
     stop_regex_strs: Optional[Union[str, List[str]]] = None  # from stop_regex
     stop_str_max_len: int = 0  # set by normalize()
     stop_regex_max_len: int = 0  # set by normalize()
+    original_temperature: Optional[float] = None
+    original_top_k: Optional[int] = None
```
Comment (rwang5203) [inline python/sglang/srt/sampling/sampling_params.py:162]: These fields shift Rust's [`is_normalized`](https://github.com/sgl-project/sglang/blob/63ccbf7a737912be4035b0286a34924756d779e7/rust/sglang-server/src/message/sampling.rs#L230-L236), so generation requests with `SGLANG_RUST_SERVER=1` fail decoding. Could you synchronize the Rust schema and capture both original sampling values before normalization?

File: docs/cookbook/autoregressive/Dream-org/Dream-v0-Base-7B.mdx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  --model-path Dream-org/Dream-v0-Base-7B \
+  --trust-remote-code \
+  --tp-size 1 \
+  --pp-size 1 \
+  --dllm-algorithm Dream \
+  --max-running-requests 4 \
+  --attention-backend flashinfer
+```
+
+Dream currently requires TP=1 and PP=1. Radix prefix caching, CUDA graphs, chunked prefill, and FDFO are disabled automatically for Dream.
```
Comment (rwang5203) [inline docs/cookbook/autoregressive/Dream-org/Dream-v0-Base-7B.mdx]: FDFO defaults to [enabled](https://github.com/sgl-project/sglang/blob/63ccbf7a737912be4035b0286a34924756d779e7/python/sglang/srt/arg_groups/fields/exec_.py#L898-L905), and NVIDIA prefill graphs remain [available](https://github.com/sgl-project/sglang/blob/63ccbf7a737912be4035b0286a34924756d779e7/python/sglang/srt/arg_groups/dllm_hook.py#L70-L80), so these limitations misdescribe the launch command. Consider documenting the actual scheduling and graph defaults.
